system/lib/pthread/proxying.c

```diff
@@ -0,0 +1,330 @@
+/*
+ * Copyright 2021 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+
+#include <assert.h>
+#include <emscripten/threading.h>
+#include <pthread.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "proxying.h"
+
+#define TASK_QUEUE_INITIAL_CAPACITY 128
+
+extern int _emscripten_notify_proxying_queue(pthread_t target_thread,
+                                             pthread_t curr_thread,
+                                             pthread_t main_thread,
+                                             em_proxying_queue* queue);
+
+typedef struct task {
+  void (*func)(void*);
+  void* arg;
+} task;
+
+// A task queue for a particular thread. Organized into a linked list of
+// task_queues for different threads.
+typedef struct task_queue {
+  // The target thread for this task_queue.
+  pthread_t thread;
+  // Recursion guard. TODO: We disallow recursive processing because that's what
+  // the old proxying API does. Experiment with relaxing this restriction.
```

Comment:
This comment reads odd. The sentence `We disallow recursive processing because that's what the old proxying API does` suggests like recursion is disallowed because the old queue did it (and it was undesired?), but then the second sentence reads like this wants to enable that as a feature?

Or does the comment want to say that "recursive processing is disallowed here, and it is not a problem since for such use cases, the old proxying API can be used instead?"

Reply:
I want to match the old behavior for now to make switching the system queue to the new queue implementation as much of a no-op as possible. Once that switch is done, it might be nice to allow recursive processing just to simplify the system. I'll make the comment more detailed.